pkg/cloudprovider/azure.go

```diff
@@ -52,9 +143,218 @@ func (p *AzureProvider) Validate(secret *kapi.Secret) []string {
 }
 
 func (p *AzureProvider) Test(secret *kapi.Secret) error {
+	var err error
+
 	if secret == nil {
 		return nil
 	}
 
+	switch p.Role {
+	case BackupStorage:
+		cloudCreds, err := godotenv.Unmarshal(string(secret.Data[AzureCredentials]))
+		if err != nil {
+			return err
+		}
+
+		blobClient, err := p.getBlobClient(cloudCreds)
+		if err != nil {
+			return err
+		}
+
+		key, _ := uuid.NewUUID()
+		test := AzureBlobTest{
+			key:       key.String(),
+			container: p.StorageContainer,
+			client:    blobClient,
+		}
+		err = test.Run()
+	}
+
+	return err
+}
+
+func (p *AzureProvider) getBlobClient(cloudCreds map[string]string) (*azstorage.BlobStorageClient, error) {
+	storageAccountKey, env, err := p.getStorageAccountKey(cloudCreds)
+	if err != nil {
+		return nil, err
+	}
+
+	storageClient, err := azstorage.NewBasicClientOnSovereignCloud(p.StorageAccount, storageAccountKey, *env)
+	if err != nil {
+		return nil, err
+	}
+
+	blobClient := storageClient.GetBlobService()
+	return &blobClient, nil
+}
+
+type AzureBlobTest struct {
+	key       string
+	container string
+	client    *azstorage.BlobStorageClient
+}
+
+func (r *AzureBlobTest) Run() error {
+	err := r.upload()
+	if err != nil {
+		return err
+	}
+	defer r.delete()
+
+	exists, err := r.objectExists()
+	if err != nil {
+		return err
+	}
+	if !exists {
+		return errors.Errorf("Azure blob key %s not found in container %s", r.key, r.container)
+	}
+
 	return nil
 }
+
+func (r *AzureBlobTest) upload() error {
+	blob, err := getBlob(*r.client, r.container, r.key)
+	if err != nil {
+		return err
+	}
+
+	blob.CreateBlockBlobFromReader(bytes.NewReader([]byte{0}), nil)
+
+	return err
+}
+
+func (r *AzureBlobTest) objectExists() (bool, error) {
```

Comment:
The other provider tests support _download()_ instead of just checking if it exists.  Why is azure different?

Reply:
My thought was that downloading the content is pointless since it isn't used, but I can see an argument in favor of downloading to ensure proper permissions are in-place (could be a different permission is needed for downloading vs. just viewing).

Reply:
Changed.